dnsdist: Add an option to forward all queries over UDP first

rgacogne opened the pull request:
### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->
This is a very rough draft!

This PR adds an option to forward queries to Do53 backends via UDP, no matter the protocol over which the query was received, and only fallback to TCP if the response is truncated. This is a big change from the current behaviour where dnsdist forwards via TCP if the query was received over Do53 TCP, DoT, DoQ or DoH3, and a lot of tests and documentation sections have to be updated. This should significantly reduce our latency for queries received over protocols other than Do53 UDP and DoH.
 
This PR also stops separating packet-cache entries depending on the protocol they were received over: we now check the size of the entry against what the client supports (looking at the EDNS UDP buffer size when available) and send a truncated answer if the cached entry is too large without having to hit the backend. This should improve our cache-hit ratio.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

rgacogne committed:
dnsdist: Add an option to forward queries over DoT via UDP

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Do not try to forward XFRs over UDP

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Only parse the query subnet for the cache once

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Stop storing separate cache entries for UDP and TCP

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Fix clang-tidy warnings

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Fix some warnings reported by clang-tidy

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Refactoring of the cache handling

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Better handling of cache lookups

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Remove useless warning

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne commented:
Rebased on current master.

pieterlexis commented on pdns/dnsdistdist/dnsdist-tcp.cc:
```diff
@@ -612,6 +689,33 @@ void IncomingTCPConnectionState::handleResponse(const struct timeval& now, TCPRe
+    dnsheader_aligned responseDH(response.d_buffer.data());
+
+    if (responseDH.get()->tc && ids.d_packet && ids.d_packet->size() > ids.d_proxyProtocolPayloadSize && ids.d_packet->size() - ids.d_proxyProtocolPayloadSize > sizeof(dnsheader)) {
+      vinfolog("Response received from backend %s via UDP, for query received from %s via TCP/DoT, is truncated, retrying over TCP", response.d_ds->getNameWithAddr(), ids.origRemote.toStringWithPort());
```
I think you'll want to use `SLOG`.

rgacogne replied:
Fixed in the meantime :+1:

pieterlexis commented on pdns/dnsdistdist/dnsdist-settings-definitions.yml:
```diff
@@ -2105,6 +2105,13 @@ general:
           Keeping ``CAP_SYS_ADMIN`` on kernel 5.8+ for example allows loading eBPF programs and altering eBPF maps at runtime even if the ``kernel.unprivileged_bpf_disabled`` sysctl is set.
           Note that this does not grant the capabilities to the process, doing so might be done by running it as root which we don't advise, or by adding capabilities via the systemd unit file, for example.
           Please also be aware that switching to a different user via ``--uid`` will still drop all capabilities."
+    - name: "forward_via_udp_first"
```
I *think* this should also be a per-frontend feature.

rgacogne replied:
Done!

rgacogne committed:
dnsdist: Fix formatting in regression tests

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne commented:
Note that this branch implements the new mode for Do 53 TCP, DoT and DoH but not DoQ/DoH3 yet (I almost forgot after working on other things for a while). I'll add DoQ and DoH3 soon.

rgacogne committed:
dnsdist: Do not try to restore the DOH unit for non-DoH queries

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Initial "forward via UDP first" support for DoQ queries

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Initial "forward via UDP first" support for DoH3 queries

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Make "forward via UDP first" a per-frontend option

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Fix ruff formatting

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne committed:
dnsdist: Don't leak the temporary object when forwarding DoQ/DoH3 queries via UDP

Signed-off-by: Remi Gacogne <remi.gacogne@powerdns.com>

rgacogne commented:
I added support for DoQ and DoH3, latency seems to be much better for DoQ, but I lack the proper tools to test DoH3.
This PR is ready for review now, I think, but the following points could use some work:
- there are a few places in the documentation where we explain how queries are forwarded based on the protocol they were received over, we need to mention the new option there
- DoTCP and DoT need the same kind of regression test I just added for DoQ and DoH3
- this can be done at a later time, but our `DOHUnitInterface` is an ugly hack that does not really make sense anymore, and with this PR we are actually extending its use. This needs to be refactored, and hopefully the `CrossProtocolQuery` interface can be used everywhere instead.

rgacogne commented:
DoQ forwarded via TCP:
<img width="1536" height="864" alt="latency" src="https://github.com/user-attachments/assets/cda6a1f1-a296-4ba8-97c5-7feaee1ab527" />

DoQ forwarded via UDP:
<img width="1536" height="864" alt="latency" src="https://github.com/user-attachments/assets/252f8d63-82f9-4875-a3a5-9e396a3b339f" />

pieterlexis commented on pdns/dnsdistdist/dnsdist-settings-definitions.yml:
```diff
@@ -2145,6 +2145,13 @@ general:
           Please also be aware that switching to a different user via ``--uid`` will still drop all capabilities."
+    - name: "forward_via_udp_first"
+      type: "bool"
+      default: "false"
```
I wonder if this should be `true` by default

pieterlexis commented on pdns/dnsdistdist/dnsdist-tcp.cc:
```diff
@@ -64,6 +64,89 @@
 
 std::atomic<uint64_t> g_tcpStatesDumpRequested{0};
 
+class TCPQueryForwardedOverUDP : public DOHUnitInterface
```
I wonder if `DOHUnitInterface` should inherit from `TCPQueryForwardedOverUDP`, as most class methods in `TCPQueryForwardedOverUDP` are there specifically for DoH (and should not be called as they throw anyway).

pieterlexis commented on pdns/dnsdistdist/dnsdist-tcp.cc:
```diff
@@ -612,6 +694,37 @@ void IncomingTCPConnectionState::handleResponse(const struct timeval& now, TCPRe
   std::shared_ptr<IncomingTCPConnectionState> state = shared_from_this();
+  auto& ids = response.d_idstate;
+  if (ids.forwardedOverUDP) {
+    dnsheader_aligned responseDH(response.d_buffer.data());
```
Should the `dnsheader_aligned` not be created *after* checking `response.d_buffer.size > sizeof(dnsheader)`?

pieterlexis commented on pdns/dnsdistdist/dnsdist-tcp.cc:
```diff
@@ -612,6 +694,37 @@ void IncomingTCPConnectionState::handleResponse(const struct timeval& now, TCPRe
+      if (g_tcpclientthreads && g_tcpclientthreads->passCrossProtocolQueryToThread(std::move(cpq))) {
+        return;
+      }
+      VERBOSESLOG(infolog("Unable to pass TCP/DoT query to a TCP worker thread after getting a TC response over UDP"),
```
Logging the client IP and qname/qtype would nice.

pieterlexis commented on pdns/dnsdistdist/doq.cc:
```diff
@@ -473,6 +596,17 @@ static void processDOQQuery(DOQUnitUniquePtr&& doqUnit)
+      // so we can use UDP to talk to the backend.
+      const dnsheader_aligned dnsHeader(unit->query.data());
+      if (dnsHeader->arcount == 0U) {
+        if (addEDNS(unit->query, 4096, false, 4096, 0)) {
```
Most nameservers use 1232 as that is considered safe over the network. This should either be configurable (e.g. when the backend is on localhost, this can be large) or be set to 1232.

pieterlexis reviewed (commented):
I gave the PR a quick look, I have some comments/questions. I'll do an in-depth review later.
